express-notes-api/server.js

```diff
@@ -0,0 +1,157 @@
+}
+
+function evaluateId(res, journal, id) {
+  const validIds = Object.keys(journal.notes);
+  const errorMsg = `${id} is not a valid ID. Please provide a valid ID: ${validIds.join(', ')}`;
+  if (isNaN(id) || Number(id) < 0) {
+    res.status(400).send(errorMsg);
+  } else if (!validIds.includes(id)) {
+    res.status(404).send(errorMsg);
+  } else {
+    return true;
+  }
+}
+
+function evaluateReqBody(req, res, action) {
+  let userInputs = Object.keys(req.body);
+  userInputs = userInputs.filter((key) => key !== 'id');
+  switch (action) {
+    case 'delete':
+      if (userInputs.length > 0) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised the above inputs were not be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+    case 'put':
+      if (!userInputs.includes('content') || userInputs.length > 1) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised only notes assigned to "content" will be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+    case 'post':
+      if (!userInputs.includes('content') || userInputs.length > 1) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised only notes assigned to "content" will be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+  }
+}
+
+/** Takes an Express response and user-provided id as inputs.
+ *  If there is an ID, return the entry for that given ID. Otherwise, return all entries. */
+async function loadNotes(res, id) {
+  try {
+    const data = await readFile('data.json', 'utf8');
+    const dataJson = JSON.parse(data);
+    if (id) {
+      if (evaluateId(res, dataJson, id)) {
+        return dataJson.notes[id];
+      }
+    } else {
+      return dataJson;
+    }
+  } catch (err) {
+    handleError(res, err);
+  }
+}
+
+async function uploadNotes(res, input) {
+  try {
+    const newNotes = JSON.stringify(input);
+    await writeFile('data.json', newNotes);
+    return true;
+  } catch (err) {
+    handleError(res, err);
+  }
+}
+
+// --- Express and middleware functions ---
+const app = express();
+
+app.use(express.json());
+
+app.get('/api/notes/:id?', async (req, res) => {
+  try {
+    const { id } = req.params;
+    let journal = null;
+    if (id) {
+      journal = await loadNotes(res, id);
+    } else {
+      journal = await loadNotes(res);
+    }
+    if (journal) {
+      const response = id ? journal : journal.notes;
+      res.status(200).send(response);
+    }
+  } catch (err) {
+    handleError(res, err);
+  }
+});
+
+app.post('/api/notes', async (req, res) => {
+  try {
+    if (evaluateReqBody(req, res, 'post')) {
+      const journal = await loadNotes(res);
+      const { nextId } = journal;
+      journal.notes[nextId] = { id: nextId, content: req.body.content };
+      journal.nextId++;
+      const response = await uploadNotes(res, journal);
```

Comment:
If you follow the advice above, you won't need to check this response. You can just rely on the exception. In other words, the line after this one (line 112) will _only_ execute if the write was successful.